### PR TITLE
osd/ECBackend: only check required shards when finishing recovery reads

### PR DESCRIPTION
**Pull request by VictorDenisov**
http://tracker.ceph.com/issues/23850

### CODE REVIEW

**Comment by smithfarm** [comment]
```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/ECBackend.cc: In member function ‘int ECBackend::get_remaining_shards(const hobject_t&, const std::set<int>&, const std::set<int>&, const ECBackend::read_result_t&, std::set<pg_shard_t>*, bool)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/ECBackend.cc:1616:55: error: no matching function for call to ‘ceph::ErasureCodeInterface::minimum_to_decode(const std::set<int>&, std::set<int>&, std::map<int, std::vector<std::pair<int, int> > >*)’
   int r = ec_impl->minimum_to_decode(want, have, &need);
                                                       ^
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/ECBackend.h:22:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/ECBackend.cc:18:
/home/jenkins-build/build/workspace/ceph-pull-requests/src/erasure-code/ErasureCodeInterface.h:286:17: note: candidate: virtual int ceph::ErasureCodeInterface::minimum_to_decode(const std::set<int>&, const std::set<int>&, std::set<int>*)
     virtual int minimum_to_decode(const std::set<int> &want_to_read,
                 ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/erasure-code/ErasureCodeInterface.h:286:17: note:   no known conversion for argument 3 from ‘std::map<int, std::vector<std::pair<int, int> > >*’ to ‘std::set<int>*’
src/osd/CMakeFiles/osd.dir/build.make:230: recipe for target 'src/osd/CMakeFiles/osd.dir/ECBackend.cc.o' failed
make[3]: *** [src/osd/CMakeFiles/osd.dir/ECBackend.cc.o] Error 1
```

**Comment by smithfarm** [comment]
@VictorDenisov Jenkins logs disappear really quickly, so I'm in the habit of pulling out (hopefully) relevant bits into comments.

Were there any conflicts when you cherry-picked the commits?

**Comment by smithfarm** [comment]
@VictorDenisov When there are conflicts, git generates a "Conflicts:" section in the commit message containing the names of the files where the conflicts occurred. Unfortunately nowadays this section is commented out by default, which is a shame because it's very useful when examining the git history to know which commits cherry-picked cleanly and which required manual conflict resolution.

Could you please re-do the cherry-picks that had conflicts and include the Conflicts section. At the very least it should show all the files where conflicts were found (as generated by git). Some folks go the extra mile and include a brief description of the conflicts and how they were resolved.

File: src/osd/ECBackend.cc [hunk 1]
```diff
@@ -1584,28 +1596,50 @@ int ECBackend::get_min_avail_to_read_shards(
...
   map<shard_id_t, pg_shard_t> shards;
+  set<pg_shard_t> error_shards;
+  for (auto &p : result.errors) {
+    error_shards.insert(p.first);
+  }
 
-  get_all_avail_shards(hoid, have, shards, for_recovery);
+  get_all_avail_shards(hoid, error_shards, have, shards, for_recovery);
+
+  map<int, vector<pair<int, int>>> need;
```
Comment (dzafman) [inline src/osd/ECBackend.cc]: @VictorDenisov This has to be adjusted for Luminous.  It should be 'set<int> need'.

File: src/osd/ECBackend.cc [hunk 2]
```diff
@@ -1584,28 +1596,50 @@ int ECBackend::get_min_avail_to_read_shards(
...
+    dout(0) << __func__ << " not enough shards left to try for " << hoid
+	    << " read result was " << result << dendl;
+    return -EIO;
+  }
 
-  for (set<int>::iterator i = have.begin();
-       i != have.end();
+  set<int> shards_left;
+  for (auto p : need) {
+    if (avail.find(p.first) == avail.end()) {
```
Comment (dzafman) [inline src/osd/ECBackend.cc]: @VictorDenisov Don't want to add _first_ since p is an stl::set not an stl::map.

File: src/osd/ECBackend.cc (hunk 1 above)
Comment (dzafman) [reply src/osd/ECBackend.cc]: Luminous doesn't include ceph/ceph#15193

**Comment by dzafman** [comment]
@smithfarm @VictorDenisov Actually, there was no code conflict.  The problem is in new lines of code that don't make sense in Luminous.  We should manually add "Conflict:" lines or at least a comment indicating changes required for Luminous.

**Comment by dzafman** [comment]
@smithfarm @VictorDenisov I'll work on the few changes that are needed.

**Comment by dzafman** [comment]
Replaced by ceph/ceph#21911

**Comment by smithfarm** [comment]
@VictorDenisov Cases like this one (cherry-picks apply without conflict, but resulting patchset is flawed) are pretty rare - don't let this deter you from backporting! ;-)

@dzafman Thanks.

**Pull request closed without merging**